### PR TITLE
ENH: one-sided HP filter implementation

### PR DESCRIPTION
**Pull request by frdrkhrng**
One-sided HP filter implementation. Test case included. Follow up of https://github.com/statsmodels/statsmodels/issues/6947

- [ ] closes #xxxx
- [ ] tests added / passed. 
- [ ] code/documentation is well formatted.  
- [ ] properly formatted commit message. See 
      [NumPy's guide](https://docs.scipy.org/doc/numpy-1.15.1/dev/gitwash/development_workflow.html#writing-the-commit-message). 

<details>

**Notes**:

* It is essential that you add a test when making code changes. Tests are not 
  needed for doc changes.
* When adding a new function, test values should usually be verified in another package (e.g., R/SAS/Stata).
* When fixing a bug, you must add a test that would produce the bug in main and
  then show that it is fixed with the new code.
* New code additions must be well formatted. Changes should pass flake8. If on Linux or OSX, you can
  verify you changes are well formatted by running 
  ```
  git diff upstream/main -u -- "*.py" | flake8 --diff --isolated
  ```
  assuming `flake8` is installed. This command is also available on Windows 
  using the Windows System for Linux once `flake8` is installed in the 
  local Linux environment. While passing this test is not required, it is good practice and it help 
  improve code quality in `statsmodels`.
* Docstring additions must render correctly, including escapes and LaTeX.

</details>

### CODE REVIEW

File: statsmodels/tsa/filters/tests/test_filters.py [hunk 1]
```diff
@@ -347,6 +347,213 @@ def test_hpfilter():
     dta = macrodata.load_pandas().data['realgdp'].values
     res = column_stack((hpfilter(dta, 1600)))
     assert_almost_equal(res, hpfilt_res, 6)
+    # Test one-sided Hodrick-Prescott Filter. Results taken from matlab.
+    hpfilt_onesided_res = array([[ 0.00000000e+00, 2.71034900e+03],
```
Comment (bashtage) [inline statsmodels/tsa/filters/tests/test_filters.py:351]: Can you seperate the test results into is own CSV file, and then load the CSV?  I'm not a big fan of large sets of values being hard coded.

File: statsmodels/tsa/filters/hp_filter.py [hunk 2]
```diff
@@ -18,6 +18,12 @@ def hpfilter(x, lamb=1600):
         suggested for quarterly data. Ravn and Uhlig suggest using a value
         of 6.25 (1600/4**4) for annual data and 129600 (1600*3**4) for monthly
         data.
+    two_sided : bool, optional
```
Comment (bashtage) [inline statsmodels/tsa/filters/hp_filter.py:21]: Maybe `one_sided=False`.  This makes it clearer that the "default" or "usual" case is 2-sdied.

File: statsmodels/tsa/filters/hp_filter.py [hunk 3]
```diff
@@ -98,4 +108,13 @@ def hpfilter(x, lamb=1600):
     trend = spsolve(I+lamb*K.T.dot(K), x, use_umfpack=use_umfpack)
 
     cycle = x - trend
+    if not two_sided:
+        one_sided_cycle = [0, 0]
+        one_sided_trend = list(x[:2])
+        for t in range(3, nobs):
+            cycle_t, trend_t = hpfilter(x[:t], lamb=lamb, two_sided=True)
```
Comment (bashtage) [inline statsmodels/tsa/filters/hp_filter.py:115]: Is this efficient?  Surely there is a better 1-sided than running `nobs` hp filters.

Comment (bashtage) [reply statsmodels/tsa/filters/hp_filter.py:115]: Looks like it should be implemented as a state-space model

https://stats.stackexchange.com/questions/48326/state-space-formulation-of-hodrick-prescott-%ef%ac%81lter/48336#48336

**Comment by josef-pkt** [comment]
Does the one-sided filter use the same weights?

To get a onesided filter, we can just take the lower triangle of `S = I+lamb*K.T.dot(K)`, that is setting the values above diagonal to zero.

I tried it out with the dense version `s2 = np.tril(S.todense())` (and rescale so that s2.sum(1) is ones), but I don't know what the expected result of the one-sided hp filter is. `np.linalg.solve(s2 / s2.sum(1)[:, None], xx)`
sparse solve is likely much faster than any statespace version

**Pull request closed without merging**